flybywiresim/a32nx
# feat(press): manual cabin pressurization

Fixes #5087 

## Summary of Changes

Second iteration of the cabin pressurization system. This one implements most of the fun bits, but it also changes some of the underlying logic of the first one, so it's important to ensure nothing has been broken on the automatic behaviour of the system. 

A few of the new modelled items and expected responses:
- When ditching overhead pushbutton is pressed, the outflow valve closes completely and packs turn off.
- When manual landing elev overhead knob is rotated so it's not in default position, the aircraft targets the manual elevation for landing (note this will create a pressure differential at touchdown!).
- When MODE SEL overhead pb is set to MAN, the outflow valve stops moving and responds to only manual inputs.
- MODE SEL MAN has precedence over Ditching (i.e. if the outflow valve is in manual, pressing the ditching button will not close it).
- Moving the MAN V/S CTL switch opens and closes the outflow valve and affects the vertical speed of the cabin.
- If cabin differential pressure goes above 8.6 PSI or below -1 PSI, the safety valves open
- The Residual Pressure Control Unit (RPCU) automatically depressurizes the aircraft in case of abnormal residual pressure on ground.
- New ECAM memos and warnings as per FCOM.

Items changed from PR #5159 :
- The system now commands a position of the outflow valve and then calculates the vertical speed of the cabin from there, much more realistic than before.
- Effects of packs on or off now should reflect on the pressurization schedule and the outflow valve opening ratio.
- The new LGCIU has been integrated.

Things not yet modelled:
- Effect of RAM AIR pushbutton on to the outflow valve.
- Electrical system integration.

## Screenshots (if necessary)


I will add a few screenshots and videos when I have this more advanced.

## References

Again, I've mainly used the FCOM and the AMM 21-30 - PRESSURIZATION CONTROL as reference. I've also had some help from a couple of active A320 pilots, but this doesn't mean everything in here is 100% accurate to real life. Any further reference or help is always welcome!!

Also a good video with good info: https://www.youtube.com/watch?v=iORXooxhYN8




## Additional context

Again I would like to thank @davidwalschots for all the help and the reviews on the first PR. I hope you can't find any of the same issues on this one 😄 .

Also a big thank you to @ParzivalSan for early testing, feedback and real life references!


Discord username (if different from GitHub): Commander Ga#1548

## Testing instructions

This should hopefully be a fun one to test. In addition to ensuring all the above items behave as advertised, here are a few suggestions to try:
1. Spawn on runway, C&D or in the air.
2. Manually close the outflow valve. 
    - If packs off, pressure should remain constant in the cabin. If in the air, a tiny amount of leakage should occur (cabin altitude should increase very slowly).
    - If packs on, pressure should build up in the cabin.
3. Manually fully open the outflow valve. 
    - If on the ground, nothing should happen.
    - If in the air, there should be rapid decompression of the cabin and a red master warning.
    - Cabin altitude should rapidly increase until it is equal to ambient.
    - Maybe follow the Excessive Cabin Altitude abnormal procedure 😉 
4. Verify cabin altitude doesn't go _above_ ambient, also that as the aircraft climbs or descends, it stays same as ambient.
5. Pressurize the cabin again by closing the outflow valve.
6. Select a manual landing elevation on the overhead different to your actual landing elevation.
7. Land with MODE MAN and validate the RPCU opens the outflow valve to depressurize the cabin.
8. Throughout the test, validate nothing "jumps", all transitions are smooth and nothing unrealistic happens.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
